### ISSUE TITLE
Chapel + LAPACK via OpenBLAS (OpenMP) Performance Issue

### ISSUE THREAD

**Issue by thomasrolinger** [chapel-lang/chapel#8337]
### Summary of Problem

I have a Chapel application that calls two LAPACK routines (`dpotrf` and `dpotrs`) one right after the other via OpenBLAS, which is parallelized with OpenMP. I noticed that the combined runtime performance of these routines are noticeably worse when compared to calling these routines from C whenever I set `OMP_NUM_THREADS` to something other than 1. I would expect to see similar performance, as both codes are calling into the same BLAS library. I have `QT_AFFINITY` set to "no", as I have heard before that this can prevent OpenMP and qthreads from stomping over each other. If I don't set this, the performance is much worse. I can get more comparable performance by explicitly setting `CHPL_RT_NUM_THREADS_PER_LOCALE` to 1 whenever I vary the number of OpenMP threads used, but the overall results are still not as good as C.

So I am wondering if there are other environment variables that need to be set to make OpenMP and Chapel/qthreads place nicely together? Or if there is some other issue present?

### Steps to Reproduce

I am attaching the full source code files for both the C and Chapel versions, as they are too long to inline. There is some set up code prior to calling the routines in question. This done to reproduce my application as much as possible, without including the entire thing. Also, I call down into the lower-level LAPACK Chapel wrappers because I am "pretending" the matrices are in column-major order to prevent any unnecessary allocation/computation to transpose a row-major matrix:

[lapack_issue.zip](https://github.com/chapel-lang/chapel/files/1682337/lapack_issue.zip)

**Compile command for Chapel code:**
`chpl -o chapel_lapack --fast -I/path/to/OpenBLAS/include -L/path/to/OpenBLAS/lib lapack_issue.chpl -lopenblas`

**Compile command for C code (-D_POSIX_C_SOURCE is for a timer in the code):**
`gcc -o c_lapack -O3 -std=c99 -D_POSIX_C_SOURCE=199309L -I/path/to/OpenBLAS/include -L/path/to/OpenBLAS/lib lapack_issue.c -lopenblas`

**Execution command for Chapel code:**

`QT_AFFINITY=no OMP_NUM_THREADS=numthreads ./chapel_lapack`

**Execution command for C code:**
`OMP_NUM_THREADS=numthreads ./c_lapack`

These are the kinds of numbers I am seeing on my system: 2x E5-2697v4 Xeon Broadwell (36 cores total). Time is in seconds and includes the time to call `dpotrf` and `dpotrs`, one right after the other, though a majority of the time is spent in `dpotrs`:

| Threads        | Chapel           | C  |
| ------------- |:-------------:| -----:|
| 1      | 0.017453  | 0.014925 |
| 2      | 0.020963  | 0.012753 |
| 4      | 0.01527    | 0.006934 |
| 8      | 0.011991  | 0.006727 |
| 16    | 0.011097  | 0.006546 |
| 32    | 0.010147  | 0.005959 |

As my real application calls these routines many times, the overall performance difference becomes much more apparent. It is worth noting that the `dpotrf` routine takes up very little of the reported times because it operates on a small matrix, so the issue can be isolated to the `dpotrs` routine

However, I can get more comparable runtimes if I explicitly set `CHPL_RT_NUM_THREADS_PER_LOCALE` to 1 for all the runs, though it's still worse than C. And clearly this isn't a fix since it will slow down the rest of my application's code.

| Threads        | Chapel  
| ------------- |:-------------:|
| 1      | 0.016391  | 
| 2      | 0.02236  | 
| 4      | 0.014037    | 
| 8      | 0.009072  | 
| 16    | 0.007901  | 
| 32    | 0.00834  | 

### Configuration Information

- Output of `chpl --version`: chpl Version 1.16.0

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads *
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc *
CHPL_MAKE: gmake
CHPL_ATOMICS: intrinsics
CHPL_GMP: gmp
CHPL_HWLOC: hwloc *
CHPL_REGEXP: re2
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-4)

**Comment by ronawho** [comment]
TL;DR: can you try setting `QT_SPINCOUNT=300` and see if that helps?

Like you mentioned, qthreads likes to assume it owns the whole machine and doesn't always play nicely with other applications.

The biggest cause of interference is from pinning qthread worker pthreads. Setting `QT_AFFINITY=no` like you're already doing should avoid that. 

The second biggest cause is from spinwaiting while looking for new tasks. Basically qthreads will create num_cores (by default, or `CHPL_RT_NUM_THREADS_PER_LOCALE` if set) pthreads at the beginning of time, and then those pthreads will spin wait waiting for a task to be queued. By default each pthread spins for 300,000 iterations before doing a condwait. While it's spinning, that can significantly degrade the performance of other threads. 300,000 iterations will run pretty fast so if you're not making any chapel tasks, the pthreads should go to sleep and not interfere, but if your application is relatively short they can interfere. Can you try setting `QT_SPINCOUNT=300` and see if that helps?

https://github.com/chapel-lang/chapel/pull/4902 and https://chapel-lang.org/releaseNotes/1.15/09-perf-opts.pdf slides 13-21 have more details or why we spinwait. 

And FYI most OpenMP implementations will also pin pthreads and spinwait (whether they do so by default is implementation specific. I think gcc does not pin by default, and has a relatively low spincount -- https://gcc.gnu.org/onlinedocs/libgomp/Environment-Variables.html `OMP_PROC_BIND` and `GOMP_SPINCOUNT` should have more details)

**Comment by thomasrolinger** [comment]
Thanks for the suggestion! My system seems to have gone down for some maintenance, so as soon as I can get on it again, I'll report back and edit this post with the result of trying `QT_SPINCOUNT=300`.

**Comment by ronawho** [comment]
FYI I finally got OpenBLAS installed. I'm running on 28 core (56 HT) broadwell nodes. I have access to a 36 core (72 HT) broadwell nodes too if we need them, but they're not availble 24/7.

For your small reproducer I'm getting pretty variable runtimes so it's hard to nail anything down. With how fast it's running though, the initial spinwaiting qthreads does could definetly impact these times. For longer running applications (over 1/10 second or so) the spincount (hopefully) shouldn't matter because the pthreads will have gone to sleep by then.

How long does your real app take to run?

**Comment by thomasrolinger** [comment]
The actual application takes a variable amount of time to run, depending on the data set. However, these routines only contribute to a small portion of that runtime.

I believe we spoke via email a couple weeks ago about this application. If you can/want, you can clone the most up-to-date version of the repository I linked you before. That has the same code that you ran but with these routines added in (i.e. they will now be the times reported for INVERSE). That way, you can get the full scope of the issue.

If you go down that path, feel free to ask here or email me with any issues with the code.
